src/library/blas3/clsparse-spm-spm.cpp

```diff
@@ -0,0 +1,1009 @@
+    // STAGE 2 - STEP 2 : create Ct
+    //cout << "nnzCt == " <<  nnzCt << endl; 
+    
+    cl_mem csrColIndCt = ::clCreateBuffer( cxt(), CL_MEM_READ_WRITE, nnzCt * sizeof( cl_int ), NULL, &run_status );
+    cl_mem csrValCt    = ::clCreateBuffer( cxt(), CL_MEM_READ_WRITE, nnzCt * sizeof( cl_float ), NULL, &run_status );
+    
+    //copy queue_one
+    run_status = clEnqueueWriteBuffer(control->queue(),
+                                     queue_one_d,
+                                     1,
+                                     0,
+                                     TUPLE_QUEUE * m * sizeof(int),
+                                     queue_one,
+                                     0,
+                                     0,
+                                     0);
+    
+    // STAGE 3 - STEP 1 : compute nnzC and Ct
+    compute_nnzC_Ct_opencl(counter_one, queue_one_d, csrRowPtrA, csrColIndA, csrValA, csrRowPtrB, csrColIndB, csrValB, csrRowPtrC, csrRowPtrCt_d, &csrColIndCt, &csrValCt, n, nnzCt, m, queue_one, control);
+    // STAGE 3 - STEP 2 : malloc C on devices
+    run_status = clEnqueueReadBuffer(control->queue(),
+                                     csrRowPtrC,
+                                     1,
+                                     0,
+                                     (m + 1)*sizeof(cl_int),
+                                     csrRowPtrC_h,
+                                     0,
+                                     0,
+                                     0);
+
+    int old_val, new_val;
+    old_val = csrRowPtrC_h[0];
+    csrRowPtrC_h[0] = 0;
+    for (int i = 1; i <= m; i++)
+    {
+        new_val = csrRowPtrC_h[i];
+        csrRowPtrC_h[i] = old_val + csrRowPtrC_h[i-1];
+        old_val = new_val;
+        //cout <<  csrRowPtrC_h[i] << " ";
+    }
+    //cout << endl;
+
+    int nnzC = csrRowPtrC_h[m];
+    //std::cout << "nnzC = " << nnzC << std::endl;
+    
+    matC->colIndices = ::clCreateBuffer( cxt(), CL_MEM_READ_WRITE, nnzC * sizeof( cl_int ), NULL, &run_status );	   
+    matC->values =     ::clCreateBuffer( cxt(), CL_MEM_READ_WRITE, nnzC * sizeof( cl_float ), NULL, &run_status );
+    
+    cl_mem csrColIndC = matC->colIndices;
+    cl_mem csrValC    = matC->values;
+
+    run_status = clEnqueueWriteBuffer(control->queue(),
+                                     csrRowPtrC,
+                                     1,
+                                     0,
+                                     (m + 1)*sizeof(cl_int),
+                                     csrRowPtrC_h,
+                                     0,
+                                     0,
+                                     0);
+
+
+    copy_Ct_to_C_opencl(counter_one, csrValC, csrRowPtrC, csrColIndC, csrValCt, csrRowPtrCt_d, csrColIndCt, queue_one_d, control);
+    
+    matC->num_rows = m;
+    matC->num_cols = n;
+    matC->num_nonzeros  = nnzC;
+
+    free(csrRowPtrC_h);
```

Comment:
All these EXPLICIT free's and clReleaseMemObjects make me nervous.  See if you can refactor with std::vectors and clsparse::vector